### PR TITLE
Process feeds all the way to the end

### PR DESCRIPTION
**Pull request by dannyroberts**
##### SUMMARY

Copying my message on slack #pillows (Mar. 11, 2020) for posterity:
> Oh! I think I figured out what the heck is going on with partitions having a medium number of items (like 500-1000) and pillows just not processing them
> Since we use python-kafka’s iterate-forever behavior, when we get to the end of the queue, until another thing makes it to that partition we will never get an opportunity to save the checkpoint
> because the kafka consumer doesn’t cede control back to our code until there’s another message to report (edited) 
> Now that we have 64 partitions and we always write to the one with the shortest backlog (on production only), the highest backlog partition will never get new messages, and so its purported backlog won’t go after it makes it’s last checkpoint and then hits the end of the queue
> when we’re not using batching, that is mainly a metrics issue: we’ve processed everything but we just never checkpoint it until another message comes through
> when you are using batching, the last few items genuinely don’t get processed, because their sitting in our buffer

### CODE REVIEW

File: corehq/ex-submodules/pillowtop/pillow/interface.py
Comment (snopoke) [inline corehq/ex-submodules/pillowtop/pillow/interface.py]: I think you can remove the next line completely since this is now called in a loop
```diff
         except PillowtopCheckpointReset:
             process_offset_chunk(changes_chunk, context)
-            self.process_changes(since=self.get_last_checkpoint_sequence(), forever=forever)
```

File: corehq/ex-submodules/pillowtop/pillow/interface.py
Comment (snopoke) [inline corehq/ex-submodules/pillowtop/pillow/interface.py]: I don't think this is ever called with `forever = True` now

File: corehq/ex-submodules/pillowtop/pillow/interface.py
Comment (dannyroberts) [reply corehq/ex-submodules/pillowtop/pillow/interface.py]: Ah thanks, yeah that makes sense

File: corehq/ex-submodules/pillowtop/pillow/interface.py
Comment (dannyroberts) [reply corehq/ex-submodules/pillowtop/pillow/interface.py]: Yeah I was also thinking that. I wanted to first see your reaction to the change before crossing the moat and pulling up the drawbridge, but it seems like you've had similar issues on ICDS and are willing to adopt this new strategy there as well (provided of course that it's well-tested)?

Comment (snopoke) [reply corehq/ex-submodules/pillowtop/pillow/interface.py]: yes, based on your description I think this is a general issue

**Comment by dannyroberts** [comment]
OK I've tested this locally and on staging with good results. I think I'm ready to put this on production. @snopoke once this PR is approved, would you be okay with me merging it and deploying to production today and making sure it works as expected? If there are any issues I would revert, and if there aren't ICDS would then get it whenever it's deployed next

File: corehq/ex-submodules/pillowtop/pillow/interface.py
Comment (millerdev) [inline corehq/ex-submodules/pillowtop/pillow/interface.py]: I know you didn't change this, but it seems strange that `process_offset_chunk()` may raise `PillowtopCheckpointReset`, and then is called in the `except` block with no error handling—is that error expected to bubble up when called there?.

**Review by millerdev — APPROVED** [review]
I don't know a lot about this part of the code base, but the theory of this change sounds good.

File: corehq/ex-submodules/pillowtop/pillow/interface.py
Comment (dannyroberts) [reply corehq/ex-submodules/pillowtop/pillow/interface.py]: I think everything you said is true; I hadn't originally noticed that this gets raised inside `process_offset_chunk`, but I think it does. That said, I don't have a good enough handle on the behavior here and what's expected to happen. I think in the worst case it just crashes and the supervisor restarts it, but I think if we were going to start messing with this we'd have to spec out the behavior we want and do more testing than just eyeballing it. @snopoke not sure if you have insights here. (I don't think it's blocking this PR, but I'm also curious.)

**Review by millerdev — APPROVED** [review]

**Comment by dannyroberts** [comment]
While testing on staging I found an issue this caused with the couch-based change feed. I'm also going to attempt to add a test or two

File: corehq/apps/change_feed/consumer/feed.py
Comment (dannyroberts) [reply corehq/apps/change_feed/consumer/feed.py]: This was more of just a logical correction covering a case that previously it wasn't thinking about, but which probably never happens in production; previously if through any combination of wiping kafka and changing checkpoints the checkpoint was ever greater than the latest offset, it would just happily keep incrementing it by one each time nonsensically. Now it moves to the latest offset. Maybe the right thing would be to have it raise an error? I can run these changes locally without this change and it doesn't seem to cause any issues. Happy to revert just this portion or revisit if you have suggestions, but I'm not sure it's worth adding a test for.

File: corehq/ex-submodules/pillowtop/pillow/interface.py
Comment (orangejenny) [inline corehq/ex-submodules/pillowtop/pillow/interface.py]: Calling this something like `last_truthy_change` would make the code read a bit more clearly and let you reduce the comments.

**Review by orangejenny — APPROVED** [review]
Approving last two changes.

File: corehq/ex-submodules/pillowtop/pillow/interface.py
Comment (dannyroberts) [reply corehq/ex-submodules/pillowtop/pillow/interface.py]: Yeah nice point. I'll try and clean this up in a followup

**Comment by dannyroberts** [comment]
Thanks Jenny!

I also wanted to just write a note about testing. I thought about what would be most useful to write tests for, I'm not sure how to approach it. The overall thing I'd love to test is that calling `run()` now does legitimately run forever and checkpoint whenever it reaches the "end" (i.e. when there are no new changes after 500 ms or something like that), but then still pick up any further changes after the pause immediately. But testing that would require multiple threads which is maybe doable with some cleverness and gevent, but also requires pausing the threads in a coordinated way to check the progress and behavior given various delays between new input etc. Any thoughts on the best way to achieve that? Beyond that, a single call to `process_changes`, which now runs until there are no new changes and then returns, is fairly well tested throughout.

I've tested pretty extensively manually locally and on staging, and it's somewhat pressing to get this out for production, so I'm going to merge and deploy and monitor how it goes. I want to be cautious even though it's tested, so I want to deploy it a low period and not on a Friday which kind of leaves now.

File: corehq/apps/change_feed/consumer/feed.py
Comment (millerdev) [reply corehq/apps/change_feed/consumer/feed.py]: My default reaction to "probably never happens in production" is not usually "not sure it's worth adding a test for." I would either determine that it _definitely never happens in production_ or write a test. I'll leave it to your judgement.

**Comment by millerdev** [comment]
> testing that would require multiple threads which is maybe doable with some cleverness and gevent

Gevent sounds like a good idea. That and liberal use of `mock.patch`.

> but also requires pausing the threads in a coordinated way to check the progress and behavior given various delays between new input etc.

Should be able to do this with `mock.patch` and gevent locks, assuming there is a method call at each point you want to pause a thread.

Granted, it's probably going to be a gnarly set of tests. One suggestion to make the tests easier to read is to build a testing framework designed around the high-level operations needed to coordinate the various moving parts in the test. Feel free to reach out if you want more detailed input.

File: corehq/apps/change_feed/consumer/feed.py
Comment (dannyroberts) [reply corehq/apps/change_feed/consumer/feed.py]: Just had a call with Cal where we traced an ICDS "pillow rewind" issue potentially back to this. (The hypothesis is that this + kafka temporarily returning 0 for partitions that are offline caused the rewind.)

I'm gonna score that as Daniel 1, Danny 0, and it's a good reminder not to throw extraneous things I don't fully understand and can't test into a larger PR.

Comment (dannyroberts) [reply corehq/apps/change_feed/consumer/feed.py]: https://github.com/dimagi/commcare-hq/pull/26899

**Pull request merged**
